hackage-security/hackage-security.cabal

```diff
@@ -149,6 +150,13 @@ library
   else
     build-depends:     base       >= 4.10
 
+  if flag(use-cabal-syntax)
+    build-depends: Cabal-syntax >= 3.7 && < 3.9
+  else
+    build-depends: Cabal        >= 1.14    && < 1.26
+                             || >= 2.0     && < 2.6
+                             || >= 3.0     && < 3.7
```

Comment:
This logic is trying to handle the split, so for:

- < 3.7, use Cabal
- >= 3.7, use Cabal-syntax

Is that the wrong version to split on?

Reply:
@phadej: ping? :)

Reply:
Oh I see, nevermind.

Reply:
> Is that the wrong version to split on?

Not really. It's a bit unfortunate as the sections are not mutually exclusive. There could be install-plan with `Cabal-syntax-3.7` and `Cabal-3.5`, so the flag value selection is arbitrary.

This could be problematic, e.g. something which uses old `Cabal` version still, then the solver will pick the `Cabal-syntax-3.7` dependency for `hackage-security`. And then weird things might happen as e.g. [`download`](https://hackage.haskell.org/package/hackage-security-0.6.0.1/docs/Hackage-Security-Client.html#v:downloadPackage) has `PackageIdentifier` in its type. Downstream would expect one from `Cabal-3.5`, but `hackage-security` would have `Cabal-syntax-3.7` one.

A solution to this is to release a dummy `Cabal-syntax-3.6` which would depend on `Cabal <3.7` and use its version to get mutual exclusive branches. The similar example is https://hackage.haskell.org/package/network-uri-2.5.0.0 but I think it's wrong as `network-uri-2.5.0.0` doesn't depend on `network` we have to write:

```
  if flag(network-uri)
    build-depends: network-uri >= 2.6, network >= 2.6
  else
    build-depends: network-uri < 2.6, network < 2.6
```

but we would like (in our case) to be able to omit the "network" from the newer branch:

```
  if flag(network-uri)
    build-depends: network-uri >= 2.6
  else
    build-depends: network-uri < 2.6, network < 2.6
```

---

I'm very tired now, please check that I didn't any mistake in my reasoning.

Reply:
... and I would rename the flag to just `Cabal-syntax`. It's common to just use the package name.

Reply:
If I understand flags correctly, assuming we have some package `foo` that depends on `Cabal-3.5` and the updated `hackage-security`, then the solver will return `Cabal-3.5` and `Cabal-syntax-3.7`. And the maintainer of `foo` would need to work around that?

Perhaps it would be more user friendly to flip the default so that `hackage-security` will use `Cabal < 3.7` and the dummy `Cabal-syntax-3.6` unless newer versions are specifically requested (`Cabal >= 3.7` or `Cabal-syntax >= 3.7`)?

In either case, what's the preferred path for generating the dummy cabal-syntax?

Reply:
If nobody objects, I will help to set that in motion.